sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala

```diff
@@ -937,4 +937,135 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+    })
+
+    // Float
+    val negativeFloat = -12345678.123f
+    val neg_floatResults: Seq[Float] =
+      Seq(-12345678f, -12345670f, -12345600f, -12345000f, -12340000f, -12300000f, -12000000f,
+        -10000000f, 0.0f, 0.0f)
+    val positiveFloat = 12345678.123f
+    val pos_FloatResults: Seq[Float] =
+      Seq(12345678f, 12345670f, 12345600f, 12345000f, 12340000f, 12300000f, 12000000f, 10000000f,
+        0.0f, 0.0f)
+
+    evaluateForEachItem(0, -10, -1)((scale, index) => {
+      checkEvaluation(TruncNumber(positiveFloat, scale), pos_FloatResults(index))
+      checkEvaluation(TruncNumber(negativeFloat, scale), neg_floatResults(index))
+    })
+
+    // Long
+    val longInput = 123456789L
+    val longResults: Seq[Long] =
+      Seq(123456789L, 123456789L, 123456789L, 123456789L, 123456780L, 123456700L, 123456000L,
+        123450000L, 123400000L, 123000000L, 120000000L, 100000000L, 0, 0)
+    evaluateForEachItem(3, -11, -1)((scale, index) => {
+      checkEvaluation(TruncNumber(longInput, scale), longResults(index))
+      checkEvaluation(TruncNumber(-longInput, scale), -longResults(index))
+    })
+
+    // Int
+    val intInput = 123456789
+    val intResults: Seq[Int] =
+      Seq(123456789, 123456789, 123456789, 123456789, 123456780, 123456700, 123456000, 123450000,
+        123400000, 123000000, 120000000, 100000000, 0, 0)
+    evaluateForEachItem(3, -11, -1)((scale, index) => {
+      checkEvaluation(TruncNumber(intInput, scale), intResults(index))
+      checkEvaluation(TruncNumber(-intInput, scale), -intResults(index))
+    })
+
+    // Short
+    val shortInput: Short = 32767
+    val shortResults: Seq[Short] =
+      Seq(32767, 32767, 32767, 32767, 32760, 32700, 32000, 30000, 0, 0)
+    evaluateForEachItem(3, -6, -1)((scale, index) => {
+      checkEvaluation(TruncNumber(shortInput, scale), shortResults(index))
+      checkEvaluation(TruncNumber(-shortInput, scale), -shortResults(index))
+    })
+
+    // Byte
+    val byteInput: Byte = 127
+    val byteResults: Seq[Byte] =
+      Seq(127, 127, 127, 127, 120, 100, 0, 0)
+    evaluateForEachItem(3, -4, -1)((scale, index) => {
+      checkEvaluation(TruncNumber(byteInput, scale), byteResults(index))
+      checkEvaluation(TruncNumber(-byteInput, scale), -byteResults(index))
+    })
+
+  checkEvaluation(TruncNumber( 12345.0, 1), 12345.0)
+  checkEvaluation(TruncNumber( Decimal(12345.1234), 1), Decimal(12345.1))
+  checkEvaluation(TruncNumber( Decimal(1.1), 1), Decimal(1.1))
+  checkEvaluation(TruncNumber( 12345.1234, 0), 12345.0)
+  checkEvaluation(TruncNumber( 12345.1234, -1), 12340.0)
+  checkEvaluation(TruncNumber( 0.1234, -1), 0.0)
+  checkEvaluation(TruncNumber( 0.1234, 0), 0.0)
+  checkEvaluation(TruncNumber( 0.0, 0), 0.0)
+  checkEvaluation(TruncNumber( 0.0, -1), 0.0)
+  checkEvaluation(TruncNumber( 0.0, 1), 0.0)
+  checkEvaluation(TruncNumber( 0.123456789, 0), 0.0)
+  checkEvaluation(TruncNumber( Decimal(0.123456789), 1), Decimal(0.1))
```

Comment:
That's what I feared. By "shrinking" the DECIMAL because we know the level of rounding we are introducing complexity/breaking changes once we decide to support genera expression rather than constants.
It woudl be more future proof to not meddle with the scale/precision here.

Reply:
Similar scale/precision change is already there in udfs `ceil` , `floor`

Reply:
Fair enough. We'll rip the bandaid for all of them at the same time.

Reply:
I've checked this behavior in Hive and presto, there also output scale/precision is changed.

Reply:
@srielau , I've modified the code to avoid changing the scale/precision of DECIMAL type.
With this new change, out DECIMAL value will retain the scale/precision of input DECIMAL value
Can you please review this PR?